sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -203,7 +204,7 @@ class Dataset[T] private[sql](
    * custom objects, e.g. collect.  Here we resolve and bind the encoder so that we can call its
    * `fromRow` method later.
    */
-  private val boundEnc =
+  private lazy val boundEnc =
```

Comment:
We can't let `boundEnc` as lazy val because we need early exception when the encoder can't be resolved.